File: java/Stacks/RepeatedRemovalOfAdjacentDuplicates.java
import java.util.Stack;

public class RepeatedRemovalOfAdjacentDuplicates {
    public String repeatedRemovalOfAdjacentDuplicates(String s) {
        Stack<Character> stack = new Stack<>();
        StringBuilder res = new StringBuilder();
        for (char c : s.toCharArray()) {
            // If the current character is the same as the top character on the stack,
            // a pair of adjacent duplicates has been formed. So, pop the top character 
            // from the stack.
            if (!stack.isEmpty() && c == stack.peek()) {
                stack.pop();
            }
            // Otherwise, push the current character onto the stack.
            else {
                stack.push(c);
            }
        }
        // Return the remaining characters as a string.
        for (char c : stack) {
            res.append(c);
        }
        return res.toString();
    }
}

File: kotlin/Stacks/RepeatedRemovalOfAdjacentDuplicates.kt
import java.util.Stack

fun repeatedRemovalOfAdjacentDuplicates(s: String): String {
    val stack = Stack<Char>()
    for (c in s) {
        // If the current character is the same as the top character on the stack,
        // a pair of adjacent duplicates has been formed. So, pop the top character
        // from the stack.
        if (stack.isNotEmpty() && c == stack.peek()) {
            stack.pop()
        }
        // Otherwise, push the current character onto the stack.
        else {
            stack.push(c)
        }
    }
    // Return the remaining characters as a string.
    return stack.joinToString("")
}
